### ISSUE TITLE
BUG: An item with the same key has already been added

### ISSUE THREAD

**Issue by hannan-azam** [OrchardCMS/Orchard#5310]
We are using a two weeks old commit from Orchard 1.9-Integration branch. We are in the process of developing public facing site of our client who is a big name in US healthcare industry. We have experienced this error three times now. Somehow duplicate content items get created in the database and the application crashes as soon as we try to go to "Content" or "Widgets".

Following is the snapshot of the error.
![error](https://cloud.githubusercontent.com/assets/12548238/7753992/83d92594-0002-11e5-81a5-ec55ad36da1f.png)

Also here is the snapshot from Audit Trail which also shows that the content item was Saved twice and then Published after which the content item got duplicated with same ID i.e. 1004.
![audittrail](https://cloud.githubusercontent.com/assets/12548238/7754300/1406c8b8-0005-11e5-922c-fe6fa0446527.png)

Last time when we faced it, client clicked on Clone link from which we felt that there could be some buggy Exporting/Importing method in one of our custom Parts which might have done this as clone does recipe export/import and we temporarily removed the Clone link from the Admin dashboard.

But this time it has happened after a Publish operation on an already existing content item which is very alarming.

In order to fix it we debug the site using VS which shows the code crash and from there we get to know the duplicated ID and then call the Delete action on it manually.

Please help.
Many Thanks

**Comment by sebastienros** [comment]
Stupid question, are you sure you are not "double-clicking" the Publish button? I could see an issue coming from that. Could you come with some repro steps?

**Comment by dcinzona** [comment]
If double clicking is the issue, there should be something in admin.js that disables the submit button when clicked.

**Comment by ivyexcellence** [comment]
@sebastienros @dcinzona we talked to the client and it seems that the double click was the culprit. There are multiple editors and someone might have clicked Save and Publish action more than once which caused this mess. We have this site hosted on Azure and sometimes the request/response doesn't get fulfilled immediately and some editor might have done double click in haste.

Moreover we did more investigation and looked into the DB. We found multiple entries for three content items in ContentItemVersionRecord table with Latest=true which again gives the clue and reconfirms what we are assuming.
![contentitemversionrecord](https://cloud.githubusercontent.com/assets/12548238/7772431/8d57d998-00b7-11e5-936c-db667af89924.png)

We are also planning to do exactly the same thing. We will disable the Save and Publish buttons once those are clicked. Page gets refreshed anyhow and user will get those enabled again once the page refresh is complete. If this gets added as out of box functionality, it will be very nice.

Regards

**Comment by sebastienros** [comment]
My suggestion is to have Orchard alter all the buttons issuing a POST to be disabled once clicked, or if possible when a form is submitted. There must be a way to intercept these events.

**Comment by remesq** [comment]
FYI, I encountered this problem with Orchard.Layouts. I accidently hit Publish twice and it threw this error both under Admin > Layouts and Admin > Content (as one content page used a layout).

@ivyexcellence gave good advice, and with it I was able to a) Go to Audit Trail (assuming you have it enabled) to determine what things were saved more than once (particularly, I needed the Content ID number and the Version number); then b) look in the database table Orchard_Frameword_ContentItemVersionRecord and find the rows that contain the Content ID and Version number and that also have "Latest" set to true.

Uncheck "Latest" on the item that was created RIGHT BEFORE the actual latest item.

**Comment by GeoffArmstrong** [comment]
I also ran into this error and fixed it by modifying the database records directly such that there is only one "latest" record for a given ContentItemRecord_id.  This post really helped me zero in on the root cause.  Thanks!

If this was a app I was working on, then I'd probably add a unique index to guarantee that there is only one "Latest" record for any given ContentItemRecord_id, if having two would break the app.  Sure, disabling the post buttons are nice, but I like having the database try to enforce data consistency.

**Comment by ivyexcellence** [comment]
@sebastienros This issue seems to be a very common one. Please assign it some milestone so that it could be added to list of bugs.

We fixed it internally by adding a small jquery snippet which ignores subsequent click events on submit.Save and submit.Publish buttons as given below. FYI we tried to disable the buttons on first click but then it wasn't even submitting the form on the first click hence we used this fix and it works like a charm.

```
<script>
    $(function () {
        var formSubmitted = false;
        $("button[name='submit.Save'], button[name='submit.Publish']").click(function (e) {
            if (formSubmitted) {
                e.preventDefault();
                return;
            }
            formSubmitted = true;
        });
    });
</script>
```

**Comment by sfmskywalker** [comment]
With the *dev* branch version this issue should no longer be there as @DanielStolt added unique key constraints. Although you would still see an error when Orchard is trying to insert duplicate rows, at least the database integrity should remain unaffected. On top of that we could intercept the click event of the save/publish buttons and disable them after the first click.

Some questions for @ivyexcellence :
1. Are you seeing these issues in a web farm environment, or on a single node?
2. After clearing out the duplicate records, are you able to reproduce the issue where duplicate records are being created again? 

I have tried clicking maniacally on Save/Publish Now from multiple browsers, but I'm only able to reproduce the issue under the following circumstance:

- SysCache enabled while running in a webfarm.

Please let me know.
Thanks!

**Comment by ivyexcellence** [comment]
@sfmskywalker We switched to 1.9.1 as soon as it was released and didn't find the fix in it so thought that it hasn't yet fixed.

Yes you are right, rather than showing an error to the user it is much better that we prevent it at UI. The small snippet if injected in the admin.js of Admin theme would handle this.

For your questions,

1. we have the site hosted on Azure with SQL instance separate from the server so yes it is a web farm environment but since we are using 1.9.1 release which probably doesn't have that fix applied and we were able to reproduce it after upgrading to 1.9.1 we went ahead with the fix.

2. Yes, after clearing the records, site comes back to normal. Since we applied that fix, we never saw this error again.

I'm not so sure about the last thing since we don't have the unique key constraint thing as we are on 1.9.1. I'll try that soon and will let you know if we are able to reproduce it.

Thanks.

**Comment by dcinzona** [comment]
That code snippet from @hannan-azam seems like it will resolve the issue.
Just wanted to point out that we experienced this problem again yesterday.
@sfmskywalker, you mentioned this is fixed in Dev?  Any chance the DB integrity check can also be pushed to 1.9.3?

**Comment by sfmskywalker** [comment]
Correct, the dev branch contains code that will create unique key constraints when setting up a new site. This means that sites running on existing databases may still experience this issue. Only recooking the site or manually creating the unique key constraints will prevent tables from receiving duplicate rows.

**Comment by dcinzona** [comment]
Thanks Sipke!  Is there a script that we can run to manually create the unique key constraints on an already deployed solution?  That may be the best option.  I can probably conjure one up but not sure what the PR was that was submitted to the dev branch that contained the unique constraints.  I'll look around, but if we can find that PR and associate it to this issue, that would help too!

**Comment by sfmskywalker** [comment]
No worries! I don;t have such a script handy unfortunately, so your best bet would indeed be to look at the updated datamigrations. The changeset you found is only part of it. There are other data migrations that add unique constraints. Perhaps the easiest way is to simply do a search on the `AddUniqueConstraint` method, which was added specifically for this fix I believe. @DanielStolt should be able to confirm; he added the code.

**Comment by DanielStolt** [comment]
@dcinzona That commit is part of it (a correction) but there should also be an earlier commit. But yes, @sfmskywalker is correct, do a search and you'll find all occurrences.

**Comment by musukvl** [comment]
I faced with this bug on 1.9.2
In my case two different users edited one document same time.
![Two latest versions](http://musuk.guru/Media/Screenshot_33.png)


Where can I find source about locking objects in Orchard?

Exception:
```
An unhandled exception has occurred and the request was terminated. Please refresh the page. If the error persists, go back
An item with the same key has already been added.
System.ArgumentException: An item with the same key has already been added. at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource) at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add) at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value) at Orchard.ContentManagement.DefaultContentManagerSession.Store(ContentItem item) at Orchard.ContentManagement.DefaultContentManager.Get(Int32 id, VersionOptions options, QueryHints hints) at Orchard.ContentManagement.DefaultContentManager.Get(Int32 id, VersionOptions options) at Orchard.ContentManagement.DefaultContentQuery.<Slice>b__26_0(ContentItemVersionRecord x) at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext() at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection) at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source) at Orchard.ContentManagement.DefaultContentQuery.Slice(Int32 skip, Int32 count) at Orchard.ContentManagement.DefaultContentQuery.ContentQuery`1.Orchard.ContentManagement.IContentQuery<T>.Slice(Int32 skip, Int32 count) at Orchard.Core.Contents.Controllers.AdminController.List(ListContentsViewModel model, PagerParameters pagerParameters) at lambda_method(Closure , ControllerBase , Object[] ) at System.Web.Mvc.ActionMethodDispatcher.Execute(ControllerBase controller, Object[] parameters) at System.Web.Mvc.ReflectedActionDescriptor.Execute(ControllerContext controllerContext, IDictionary`2 parameters) at System.Web.Mvc.ControllerActionInvoker.InvokeActionMethod(ControllerContext controllerContext, ActionDescriptor actionDescriptor, IDictionary`2 parameters) at System.Web.Mvc.Async.AsyncControllerActionInvoker.<BeginInvokeSynchronousActionMethod>b__39(IAsyncResult asyncResult, ActionInvocation innerInvokeState) at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncResult`2.CallEndDelegate(IAsyncResult asyncResult) at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncResultBase`1.End() at System.Web.Mvc.Async.AsyncControllerActionInvoker.EndInvokeActionMethod(IAsyncResult asyncResult) at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3d() at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<>c__DisplayClass46.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3f() at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<>c__DisplayClass46.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3f() at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<>c__DisplayClass46.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3f() at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<>c__DisplayClass46.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3f() at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<>c__DisplayClass46.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3f()
 ```

**Comment by sfmskywalker** [comment]
The way this is fixed in the *dev* branch is by adding unique key constraints to this and other tables. That won't of course fix your existing data, but once the duplicates have been removed, it will at least prevent duplicate records in the future.

**Comment by m8ker** [comment]
I'm getting this error on 1.9.1. I went into the database to look for ContentItemRecord_id duplicates in Orchard_Framework_ContentItemVersionRecord that are both set to latest="true" and didn't find any. I did find sets of duplicates where the latest record was set to "false" along with the other duplicates. When I set the most recent duplicates to "true" that didn't help either. 

Is there another bug that could be causing this? Any tips on how to "hard delete" these items or any other way to fix/work around the issue? If I can't get into the content items area, I can't delete the items. 

```
An unhandled exception has occurred and the request was terminated. Please refresh the page. If the error persists, go back
An item with the same key has already been added.
System.ArgumentException: An item with the same key has already been added. at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource) at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add) at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value) at Orchard.ContentManagement.DefaultContentManagerSession.Store(ContentItem item) in D:\home\site\repository\src\Orchard\ContentManagement\DefaultContentManagerSession.cs:line 12 at Orchard.ContentManagement.DefaultContentManager.Get(Int32 id, VersionOptions options, QueryHints hints) in D:\home\site\repository\src\Orchard\ContentManagement\DefaultContentManager.cs:line 231 at Orchard.ContentManagement.DefaultContentManager.Get(Int32 id, VersionOptions options) in D:\home\site\repository\src\Orchard\ContentManagement\DefaultContentManager.cs:line 131 at Orchard.ContentManagement.DefaultContentQuery.<Slice>b__26_0(ContentItemVersionRecord x) in D:\home\site\repository\src\Orchard\ContentManagement\DefaultContentQuery.cs:line 193 at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext() at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection) at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source) at Orchard.ContentManagement.DefaultContentQuery.Slice(Int32 skip, Int32 count) in D:\home\site\repository\src\Orchard\ContentManagement\DefaultContentQuery.cs:line 191 at Orchard.ContentManagement.DefaultContentQuery.ContentQuery`1.Orchard.ContentManagement.IContentQuery<T>.Slice(Int32 skip, Int32 count) in D:\home\site\repository\src\Orchard\ContentManagement\DefaultContentQuery.cs:line 296 at Orchard.Core.Contents.Controllers.AdminController.List(ListContentsViewModel model, PagerParameters pagerParameters) in D:\home\site\repository\src\Orchard.Web\Core\Contents\Controllers\AdminController.cs:line 126 at lambda_method(Closure , ControllerBase , Object[] ) at System.Web.Mvc.ActionMethodDispatcher.Execute(ControllerBase controller, Object[] parameters) at System.Web.Mvc.ReflectedActionDescriptor.Execute(ControllerContext controllerContext, IDictionary`2 parameters) at System.Web.Mvc.ControllerActionInvoker.InvokeActionMethod(ControllerContext controllerContext, ActionDescriptor actionDescriptor, IDictionary`2 parameters) at System.Web.Mvc.Async.AsyncControllerActionInvoker.<BeginInvokeSynchronousActionMethod>b__39(IAsyncResult asyncResult, ActionInvocation innerInvokeState) at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncResult`2.CallEndDelegate(IAsyncResult asyncResult) at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncResultBase`1.End() at System.Web.Mvc.Async.AsyncControllerActionInvoker.EndInvokeActionMethod(IAsyncResult asyncResult) at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3d() at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<>c__DisplayClass46.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3f() at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<>c__DisplayClass46.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3f() at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<>c__DisplayClass46.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3f()
```

**Comment by sfmskywalker** [comment]
The only way I know of is by setting the **Latest** and **Published** columns to false for the duplicate version records. Except of course for the record that is actually the latest and the one that is actually published. One record can be both Latest and Published, but no two records should be Published or Latest at the same time, as you unfortunately found out.

As of 1.10 (and perhaps even 1.9.3), this issue is prevented by means of unique key constraints.

**Comment by m8ker** [comment]
Thanks sfmskywalker. I went through and made sure "Latest" and "Published" are both false. The table looks right, but I'm still not able to view content items in the CMS. I'm not sure what to do at this point.
 
Delete the problematic content type altogether and re-make it with a different name? 

Or update Orchard and export/import the old data? If I updated Orchard and imported the old data, would the problem persist?

**Comment by sfmskywalker** [comment]
Perhaps also make sure to recycle the application pool, maybe **SysCache** is keeping these items in memory?

I ran into this very same issue, but was able to solve it after carefully identifying all duplicate version records. I missed a fee instances on occasion, it's easy to overlook one.

Updating Orchard won't solve your issue of duplicate records unfortunately; instead, you would most likely start seeing database migration errors due to the unique key constraints trying to be applied, which would fail due to duplicate records.

I wrote a SQL query at the time to select all duplicates (there were quite a few). Unfortunately I don;t have that anymore, but I highly recommend writing one yourself. The following SO post helped me write one: http://stackoverflow.com/questions/2594829/finding-duplicate-values-in-a-sql-table
